### PR TITLE
NXP Backend: Add imxrt700cm backend which combines the Neutron and CortexM backends

### PR DESCRIPTION
**Pull request by MartinPavella**
### Summary
Add imxrt700cm backend which combines the Neutron and CortexM backends into one. The backend uses Neutron wherever possible, and the leftover nodes are handled by Cortex-M.

Fixes #13470

### Test plan
Unit tests provided

cc @robert-kalmar @JakeStevens @digantdesai @rascani

### CODE REVIEW

File: examples/nxp/experimental/cifar_net/cifar_net.py [hunk 1]
```diff
@@ -137,9 +162,11 @@ def get_model(
                              training, the weights will be stored in the file.
     :param train: Boolean indicating whether to train the model.
     :param num_epochs: Number of epochs to use during training.
+    :param cortex_m_safe: There is a bug in the Cortex-M backend related to the `pad` operator. If this parameter is
```
Comment (digantdesai) [inline examples/nxp/experimental/cifar_net/cifar_net.py]: Let's fix this if it is something quick as opposed to introducing a new bypass logic? WDYT - CC @rascani since you were discussing this earlier this week in the context of NHWC.

File: backends/nxp/imxrt700cm/imxrt700cm_quantizer.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+from executorch.backends.cortex_m.quantizer.quantizer import CortexMQuantizer
+from executorch.backends.nxp import NXP_NEUTRON_BACKEND_IGNORE
+from executorch.backends.nxp.backend.neutron_target_spec import NeutronTargetSpec
+from executorch.backends.nxp.quantizer.neutron_quantizer import NeutronQuantizer
+from torch.fx import GraphModule, Node
+from torchao.quantization.pt2e.quantizer import Quantizer
+from torchao.quantization.pt2e.quantizer.quantizer import Q_ANNOTATION_KEY
+
+
+class IMXRT700CMQuantizer(Quantizer):
```
Comment (digantdesai) [inline backends/nxp/imxrt700cm/imxrt700cm_quantizer.py:17]: Quantizers are meant to be composible. Recipe is the right user facing abstraction to target an SoC with multiple different backends. Take a look at https://github.com/pytorch/executorch/blob/main/export/tests/test_target_recipes.py especially something like `get_android_recipe` to understand how two or more quantizers / partitioners are encapsulated and made to work together.

In your case, I imagine a target recipe for rt700 with neutron and cortex-m.

**Review by digantdesai — CHANGES_REQUESTED** [review]

File: examples/nxp/experimental/cifar_net/cifar_net.py [hunk 3]
```diff
@@ -27,14 +27,28 @@
...
-    def __init__(self, batch_size: int = 1, pth_file: str | None = None):
+    def __init__(
+        self,
+        batch_size: int = 1,
+        pth_file: str | None = None,
+        cortex_m_safe: bool = False,
+    ):
+        """
+        :param cortex_m_safe: There is a bug in the Cortex-M backend related to the `pad` operator. If this parameter is
+                               `True`, padding will be done via `Conv2d` attribute `padding` instead of `F.pad`.
```
Comment (robert-kalmar) [inline examples/nxp/experimental/cifar_net/cifar_net.py]: FYI, time to address the https://github.com/pytorch/executorch/issues/13470 to simplify the CifarNet model.

File: backends/nxp/imxrt700cm/imxrt700cm_quantizer.py (hunk 2 above)
Comment (MartinPavella) [reply backends/nxp/imxrt700cm/imxrt700cm_quantizer.py:17]: Thank you @digantdesai  for the insights. I have looked into it, and recipes definitely look like the right way forward.  
I analyzed the state in executorch: 

- To introduce an SoC recipe would require having recipes for both Neutron and Cortex-M backend (both missing). Alternatively the current Cortex-M and Neutron pipelines can be combined into a single recipe but from reuse perspective a base recipe for both backend seems better from my opinion. Our Neutron backend pipeline is currently implemented in https://github.com/pytorch/executorch/blob/28b48132eef3d6655b68801c91fae4a100f5992a/backends/nxp/tests/executorch_pipeline.py#L117

- The Neutron pipeline contains some kernel registration functionality, as only it knows what NPU kernels are requires. This would probably require the creation of a new `Stage` type https://github.com/pytorch/executorch/blob/28b48132eef3d6655b68801c91fae4a100f5992a/backends/nxp/tests/executorch_pipeline.py#L117
Or at least I didn't find any stage providing the functionality to just execute a function based on presence of an option. 

- The QAT appears to not be supported. The `QuantizeStage` explicitly states it performs post-training quantization. I see that the `SourceTransformStage` also enables quantization in some way, but it doesn't seem QAT is supported. So perhaps this would require another new Stage type (or modification on an existing stage).

 
Given this, enabling the RT700 Neutron+Cortex-M backend via a recipe requires changes in multiple backends, and this PR would end up quite large. Can we do this in multiple stages? Such as:
 
1. Experimentally, continue with this early implementation introducing the option to combine Cortex-M and Neutron Backends for the i.MXRT700.
2. Rework the current Neutron lowering pipeline to a recipe, and the same for the Cortex-M backend. Here we would potentially introduce new Stages.
3. Rework the `imxrt700cm` lowering to a recipe
4. Based on consequent discussion extend for QAT training. 
 
For Cortex-M we need to sync with Arm too. 
What is your opinion?

File: examples/nxp/experimental/cifar_net/cifar_net.py (hunk 1 above)
Comment (MartinPavella) [reply examples/nxp/experimental/cifar_net/cifar_net.py]: Good point.

The issue with the `pad` operator is that calling the pad replacement op in https://github.com/pytorch/executorch/blob/28b48132eef3d6655b68801c91fae4a100f5992a/backends/cortex_m/passes/quantized_op_fusion_pass.py#L410-L415
would produce a contiguous output even when the input had the channels last dim order. I tried to look for the root cause but I didn't find it, so I opted for the bypass and I planned to report the issue after raising this PR.

File: examples/nxp/experimental/cifar_net/cifar_net.py (hunk 3 above)
Comment (MartinPavella) [reply examples/nxp/experimental/cifar_net/cifar_net.py]: Thanks for letting me know about this issue. I will resolve it as part of this PR.

File: examples/nxp/experimental/cifar_net/cifar_net.py (hunk 1 above)
Comment (rascani) [reply examples/nxp/experimental/cifar_net/cifar_net.py]: I have a fix for pad here: https://github.com/pytorch/executorch/pull/18429

Comment (MartinPavella) [reply examples/nxp/experimental/cifar_net/cifar_net.py]: Thank you!
I have removed the workaround from our testing model.

File: backends/nxp/imxrt700cm/imxrt700cm_quantizer.py (hunk 2 above)
Comment (digantdesai) [reply backends/nxp/imxrt700cm/imxrt700cm_quantizer.py:17]: Hey @MartinPavella 

> To introduce an SoC recipe would require having recipes for both Neutron and Cortex-M backend (both missing)

Well that would be nice to have, no? :)

> The Neutron pipeline contains some kernel registration functionality, as only it knows what NPU kernels are requires. This would probably require the creation of a new Stage type 

Are you referring to `handle_kernel_selection`? (I haven't read the code for this selective build so correct me if I am wrong). I would argue the need for this to be visible at all outside the delegate in the lowering flow. But I don't want to block you here, adding a new arg for a pass or a fn for ToBackendStage might not be a bad choice.

The idea for the recipe is to conform to a standard high level flow to make your delegate composable with other delegates. So you arguing for a special path is opposite of what a composable delegate would do ;)

> The QAT appears to not be supported.

As you alluded, does QAT have to be supported from the get go? I would say feel free to update the Quant stage to support that.

> Can we do this in multiple stages? 

Sounds good. I don't want us to go down (1) too deep and ignore the recipes once we have (1) working. What do you have a timeline for removing (1) (assuming we do land this PR for time being).

Also @rascani can support you for Cortex-M, he is the lead for Cortex-M.

Comment (robert-kalmar) [reply backends/nxp/imxrt700cm/imxrt700cm_quantizer.py:17]: Hey @digantdesai, 
1. Not just nice to have. We intend to leverage the recipes :-). As you pointed out the recipe concept, Martin looked on it, what it involves and listed the items needed. They are already in our backlog. We plan to introduce recipes for Neutron, Cortex-M, Cadence, either by us or with the impacted backends leads. 
2. `handle_kernel_selection`, it actually can be (and should) part of the `compile_spec`, or alternatively the kernel selection code can be dumped by default. The thing is the compiler sees each delegated partition separately and for each lists the kernels present in the partition. 
And finally, one needs to merge those partial kernels list into one final set.
What I can imagine is an additional step in the `LoweringRecipe` defining transformations or post processing functions applied after all the to_backends calls. 
Alternatively it can be done manually by a user itself too. 
Let us think about it, and come up with some proposal. 
3. QAT - No, not needed immediately. Another backlog item listed by Martin. 
4. Timeline: For ExecuTorch 1.3 (~May/2026) we already have items on the roadmap related to increasing the operator support. The recipe is a backlog item to be addressed based on our capacity (for ET 1.3). For ET 1.4 (Jul) - 1.5 (Sep) it is on our roadmap. 
Note: I derived the months based on current ExecuTorch release cadence and PyTorch releases :-)
So roughly Q2 or early Q3.
